src/ledger/backend/ethereumOnChainQuerier.spec.ts

```diff
@@ -0,0 +1,110 @@
+"use strict";
+
+import { expect } from "chai";
+import { EthereumOnChainQuerier } from "./ethereumOnChainQuerier";
+import { OnChainQuerier } from "../onChainQuerier";
+
+import * as test from "#erdstall/test";
+import { Environment, setupEnv } from "test/ledger";
+import { Wallet } from "@ethersproject/wallet";
+import { IERC721Minter__factory, IERC721__factory } from ".";
+import { Tokens } from "../assets";
+import PRNG from "#erdstall/test/random";
+import { afterEach } from "mocha";
+
+const PRIMES = [2, 3, 5, 7];
+const TOKEN_SIZE = PRIMES.length;
+const RNG: PRNG = test.newPrng();
+
+interface Entity {
+	wallet: Wallet;
+	tokens: Tokens;
+	runningTokens: bigint[];
+}
+
+describe("EthereumOnChainQuerier", function () {
+	let testenv: Environment;
+	let alice: Entity;
+	let bob: Entity;
+	let onChainQueryA: OnChainQuerier;
+	let aTransferSeed: number;
+	let bTransferSeed: number;
+
+	const initEntity = async (wallet: Wallet): Promise<Entity> => {
+		const entity = {} as Partial<Entity>;
+		entity.wallet = wallet;
+		entity.tokens = test.newRandomTokens(RNG, TOKEN_SIZE);
+		entity.runningTokens = [];
+		for (const token of entity.tokens.value) {
+			entity.runningTokens.push(token);
+		}
+
+		const erc721Minter = IERC721Minter__factory.connect(
+			testenv.perunArt,
+			entity.wallet,
+		);
+		for (const id of entity.tokens.value) {
+			await erc721Minter.mint(entity.wallet.address, id);
+		}
+
+		return entity as Entity;
+	};
+
+	const seededTokenTransfer = async (
+		from: Entity,
+		to: Entity,
+		seed: number,
+		tokenPool: Tokens,
+	): Promise<void> => {
+		const ierc721 = IERC721__factory.connect(testenv.perunArt, from.wallet);
+		for (let index = 0; index < TOKEN_SIZE; ++index) {
```

Comment:
Consider using `PRIMES.forEach((p, i) => {...})`.
I'd also define `const tknId = tokenPool.value[i]` because you use this expression three times inside the loop.

Reply:
I got used to not using `forEach` while developing JS backend because of the performance implications. In case you ever test hundreds of thousands of transactions: You have been warned :warning:

Reply:
And our test env doesn't play nice with multiple async transfer calls, so this won't work either.